### ISSUE TITLE
Model compression problem

### ISSUE THREAD

**Issue by ewwll** [keras-team/keras#14386]
I now compress a model, but why is the compressed model bigger than before? I use the official tutorial example
class Distiller(keras.Model):
    def __init__(self, student, teacher):
        super(Distiller, self).__init__()
        self.teacher = teacher
        self.student = student

    def compile(
        self,
        optimizer,
        metrics,
        student_loss_fn,
        distillation_loss_fn,
        alpha=0.1,
        temperature=3,
    ):
        """ Configure the distiller.

        Args:
            optimizer: Keras optimizer for the student weights
            metrics: Keras metrics for evaluation
            student_loss_fn: Loss function of difference between student
                predictions and ground-truth
            distillation_loss_fn: Loss function of difference between soft
                student predictions and soft teacher predictions
            alpha: weight to student_loss_fn and 1-alpha to distillation_loss_fn
            temperature: Temperature for softening probability distributions.
                Larger temperature gives softer distributions.
        """
        super(Distiller, self).compile(optimizer=optimizer, metrics=metrics)
        self.student_loss_fn = student_loss_fn
        self.distillation_loss_fn = distillation_loss_fn
        self.alpha = alpha
        self.temperature = temperature

    def train_step(self, data):
        # Unpack data
        x, y = data

        # Forward pass of teacher
        teacher_predictions = self.teacher(x, training=False)

        with tf.GradientTape() as tape:
            # Forward pass of student
            student_predictions = self.student(x, training=True)

            # Compute losses
            student_loss = self.student_loss_fn(y, student_predictions)
            distillation_loss = self.distillation_loss_fn(
                tf.nn.softmax(teacher_predictions / self.temperature, axis=1),
                tf.nn.softmax(student_predictions / self.temperature, axis=1),
            )
            loss = self.alpha * student_loss + (1 - self.alpha) * distillation_loss

        # Compute gradients
        trainable_vars = self.student.trainable_variables
        gradients = tape.gradient(loss, trainable_vars)

        # Update weights
        self.optimizer.apply_gradients(zip(gradients, trainable_vars))

        # Update the metrics configured in `compile()`.
        self.compiled_metrics.update_state(y, student_predictions)

        # Return a dict of performance
        results = {m.name: m.result() for m in self.metrics}
        results.update(
            {"student_loss": student_loss, "distillation_loss": distillation_loss}
        )
        return results
    def save_Model(self):
        self.student.save("student.json")
        self.teacher.save("teacher.json")
    
    def save_Weight(self):
        self.student.save_weights("student_weight.h5")
        self.teacher.save_weights("teacher_weight.h5")

    def test_step(self, data):
        # Unpack the data
        x, y = data

        # Compute predictions
        y_prediction = self.student(x, training=False)

        # Calculate the loss
        student_loss = self.student_loss_fn(y, y_prediction)

        # Update the metrics.
        self.compiled_metrics.update_state(y, y_prediction)

        # Return a dict of performance
        results = {m.name: m.result() for m in self.metrics}
        results.update({"student_loss": student_loss})
        return results
    def call(self, inputs):
        return self.student(inputs, training=False)

# # Create the teacher
# teacher = Sequential()
# # model.add(ResNet50(include_top=False,weights=None,input_shape=(480,640,1),classes=9))
# teacher.add(Conv2D(filters= 32, kernel_size=(6,6), padding='valid',input_shape=(480,640,1)))
# teacher.add(LeakyReLU(alpha=0.05))
# teacher.add(MaxPool2D(pool_size=(2,2)))
# teacher.add(Dropout(0.25))
# teacher.add(Conv2D(filters= 16, kernel_size=(5,5), padding='valid'))
# teacher.add(LeakyReLU(alpha=0.05))
# teacher.add(MaxPool2D(pool_size=(2,2)))
# teacher.add(Dropout(0.25))
# teacher.add(Conv2D(filters= 8, kernel_size=(4,4), padding='valid'))
# teacher.add(LeakyReLU(alpha=0.05))
# teacher.add(MaxPool2D(pool_size=(2,2)))
# teacher.add(Dropout(0.25))
# teacher.add(Conv2D(filters= 4, kernel_size=(3,3), padding='valid',strides=2))
# teacher.add(LeakyReLU(alpha=0.05))
# teacher.add(MaxPool2D(pool_size=(2,2)))
# teacher.add(Dropout(0.25))
# teacher.add(Conv2D(filters= 2, kernel_size=(2,2), padding='valid',strides=2))
# teacher.add(LeakyReLU(alpha=0.05))
# teacher.add(MaxPool2D(pool_size=(2,2),strides=2))
# teacher.add(Dropout(0.25))
# teacher.add(Flatten())
# teacher.add(Dense(256, activation='sigmoid'))
# teacher.add(BatchNormalization())
# teacher.add(Dense(len(type_name), activation='softmax'))

if forwhat_num==0:
    teacher=load_model("./model_bak/letter_acc93_epoch8.h5")
elif forwhat_num==1:
    teacher=load_model("./model_bak/weight-number-07-0.95.h5")
else:
    teacher=load_model("./model_bak/symbol_acc96_epoch15.h5")

# Create the student
student = Sequential()
student.add(Flatten(input_shape=(480,640,1)))
student.add(Dense(128, activation='relu'))
student.add(Dense(len(type_name), activation='softmax'))

# Clone student for later comparison
student_scratch = keras.models.clone_model(student)

distiller = Distiller(student=student, teacher=teacher)
distiller.compile(
    optimizer=keras.optimizers.Adam(),
    metrics=[keras.metrics.SparseCategoricalAccuracy()],
    student_loss_fn=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
    distillation_loss_fn=keras.losses.KLDivergence(),
    alpha=0.1,
    temperature=10,
)

checkpoint_path = "./weight-distiller-"+forwhat[forwhat_num]+".h5"
cp_callback = tf.keras.callbacks.ModelCheckpoint(filepath=checkpoint_path, save_weights_only=False, monitor='val_acc', save_best_only=True, verbose=1)

callbacks_list = [cp_callback]

# Distill teacher to student
distiller.fit(imgall, label, epochs=10,validation_data=(testimgall,testlabel),callbacks=callbacks_list)
distiller.evaluate(testimgall, testlabel)
student.save("my_model.h5")
student.save_weights("my_wight.h5")
distiller.save_Model()
distiller.save_Weight()
distiller.save_weights("d.h5")
# Evaluate student on test dataset
tf.saved_model.save(distiller,"distiller_{}_graph".format(forwhat[forwhat_num]))
pres=distiller(testimgall,training=False)
out=[]
m=np.size(pres, axis=0)
pres=pres.tolist()
print(pres)

for i in range(m):
    item=pres[i]
    index=item.index(np.max(pres))
    out.append(index)
print(out)
Why is the compressed model bigger than before?

**Comment by ariG23498** [comment]
Hey @ewwll could you share a colab instead of the what you have given here! Right now it is not readable or executable. After you share the code, I can reproduce and check for any issues.

**Comment by ewwll** [comment]
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import sys
import os
import cv2 
import numpy as np
import tensorflow as tf
from keras.callbacks import ModelCheckpoint
import matplotlib.pyplot as plt
from keras.models import *
from keras.layers import *
from keras.optimizers import *
from keras.callbacks import ModelCheckpoint, LearningRateScheduler
from keras import backend as keras
from keras import applications
trainname=os.listdir("./camera/image/")
testname=os.listdir("./camera/label/")
print(trainname)
print(len(trainname))
print(testname)
print(len(testname))
train=[]
test=[]
train_label=[]
test_label=[]
n=0
for i in trainname:
    if(n<200):
        img=cv2.imread("./camera/image/"+i)
        img=cv2.resize(img,(252,189))
        img=cv2.cvtColor(img,cv2.COLOR_RGB2GRAY)
        img=img/255
        train.append(img)
        j=0
        num=[]
        while(j<6):
            img1=cv2.imread("./camera/label/"+str(n)+'_'+str(j)+".jpg")
            img1=cv2.resize(img1,(252,180))
            img1=cv2.cvtColor(img1,cv2.COLOR_RGB2GRAY)
            img1=img1/255
            img1=img1.transpose()
            num.append(img1)
            j=j+1
        num=np.array(num)
        num=num.transpose()
        train_label.append(num)
    else:
        img=cv2.imread("./camera/image/"+i)
        img=cv2.resize(img,(252,189))
        img=cv2.cvtColor(img,cv2.COLOR_RGB2GRAY)
        img=img/255
        test.append(img)
        j=0
        num=[]
        while(j<6):
            img1=cv2.imread("./camera/label/"+str(n)+'_'+str(j)+".jpg")
            img1=cv2.resize(img1,(252,180))
            img1=cv2.cvtColor(img1,cv2.COLOR_RGB2GRAY)
            img1=img1/255
            img1=img1.transpose()
            num.append(img1)
            j=j+1
        print(len(num))
        num=np.array(num)
        num=num.transpose()
        print(num.shape)
        test_label.append(num)
    n=n+1
plt.imshow(train[0])
plt.show()
plt.imshow(test[0])
plt.show()
print(train_label[0].shape)
project_name = "fcn_segment"
channels = 1
std_shape = (189, 252, channels)
model = Sequential(name = project_name)
# model.add(Conv2D(32, kernel_size = (3, 3), activation = "relu",padding = "same", input_shape = std_shape,name = "conv_1"))
# model.add(MaxPool2D(pool_size = (2, 2), strides = (2, 2), name = "max_pool_1"))
# model.add(Dropout(0.25))
# model.add(Conv2D(64, kernel_size = (3, 3), activation = "relu",padding = "same", name = "conv_2"))
# model.add(MaxPool2D(pool_size = (2, 2), strides = (2, 2), name = "max_pool_2"))
# model.add(Dropout(0.25))
# model.add(Conv2D(128, kernel_size = (3, 3), activation = "relu",padding = "same", name = "conv_3"))
# model.add(MaxPool2D(pool_size = (2, 2), strides = (2, 2), name = "max_pool_3"))
# model.add(Dropout(0.25))
# model.add(Conv2D(256, kernel_size = (3, 3), activation = "relu",padding = "same", name = "conv_4"))
# model.add(MaxPool2D(pool_size = (2, 2), strides = (2, 2), name = "max_pool_4"))
# model.add(Dropout(0.25))
# model.add(Conv2D(512, kernel_size = (3, 3), activation = "relu",padding = "same", name = "conv_5"))
# model.add(MaxPool2D(pool_size = (2, 2), strides = (2, 2), name = "max_pool_5"))
# model.add(Dropout(0.25))
model.add(applications.VGG19(input_shape=std_shape,include_top=None,weights=None))
model.add(BatchNormalization())
model.add(UpSampling2D(size = (6, 6), interpolation = "nearest", name = "upsamping_6"))
model.add(Conv2D(3, kernel_size = (5, 5), activation = "sigmoid",padding = "same", name = "conv_6"))
model.add(UpSampling2D(size = (6, 6), interpolation = "nearest", name = "upsamping_7"))
model.add(Conv2D(6, kernel_size = (3, 3), activation = "sigmoid",padding = "same", name = "conv_7"))
model.compile(optimizer = "adam",loss = "mean_squared_error",metrics = ["accuracy"])
model.summary()
train=np.array(train)
test=np.array(test)
train_label=np.array(train_label)
test_label=np.array(test_label)
print(train.shape)
print(test.shape)
print(train_label.shape)
print(test_label.shape)
train=np.expand_dims(train,axis=3)
test=np.expand_dims(test,axis=3)
print(train.shape)
print(test.shape)
filepath = "best_weights.h5"
checkpoint = ModelCheckpoint(filepath, monitor='accuracy', verbose=0, save_best_only=True, mode='max')
callbacks_list = [checkpoint]
history = model.fit(train, train_label,batch_size =6,epochs = 40, shuffle=True,verbose = 1,validation_data=(test,test_label),callbacks = callbacks_list)
acc = history.history['accuracy']
loss = history.history['loss']
epochs = range(1, len(acc) + 1)
plt.title('Accuracy and Loss')
plt.plot(epochs, acc, 'red', label='Training acc')
plt.plot(epochs, loss, 'blue', label='Validation loss')
plt.legend()
plt.show()

**Comment by gadagashwini** [comment]
@ewwll,
Create 3x smaller models from pruning, you can use pruning method to compress the model. 
For more information read [here](https://www.tensorflow.org/model_optimization/guide/pruning/pruning_with_keras#create_3x_smaller_models_from_pruning)
